libzecale/circuits/aggregator_gadget.hpp

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZECALE_CIRCUITS_AGGREGATOR_GADGET_HPP_
+#define __ZECALE_CIRCUITS_AGGREGATOR_GADGET_HPP_
+
+#include <libff/algebra/fields/field_utils.hpp>
+#include <libsnark/gadgetlib1/gadget.hpp>
+#include <libsnark/gadgetlib1/gadgets/basic_gadgets.hpp>
+#include <libzeth/core/extended_proof.hpp>
+
+namespace libzecale
+{
+
+/// Gadget that aggregates a batch of nested proofs, creating a single wrapping
+/// proof the validity (or possible invalidity) of each proof in the batch.
+///
+/// A note about fields:
+/// A proof (PGHR13 or GROTH16) is made of group elements
+/// (G1 or G2 for some pairing) where the group elements belong to E/F_q (for
+/// G1) or E/F_q^n (for G2), and `n` varies depending on the setting. As such,
+/// the coordinates of the elements in the proof are defined over the field
+/// F_q. This field is referred to as the "base field". Primary inputs however
+/// are defined over F_r, referred to as the "scalar field".
+///
+/// In order to aggregate proofs, we require that the base field of the curve
+/// used in the nested proof (nppT here) be the scalar field for the wrapping
+/// pairing (wppT).
+template<
+    typename nppT,
+    typename wppT,
+    typename nsnarkT,
+    typename wverifierT,
+    size_t NumProofs>
+class aggregator_gadget : libsnark::gadget<libff::Fr<wppT>>
+{
+private:
+    using verifier_gadget = typename wverifierT::verifier_gadget;
+    using proof_variable_gadget = typename wverifierT::proof_variable_gadget;
+    using verification_key_variable_gadget =
+        typename wverifierT::verification_key_variable_gadget;
+    using input_packing_gadget = libsnark::multipacking_gadget<libff::Fr<wppT>>;
+
+    const size_t num_inputs_per_nested_proof;
+
+    /// The nested primary inputs lie in the scalar field `libff::Fr<nppT>`,
+    /// and must be represented as elements of `libff::Fr<wppT>` for use in the
+    /// wrapper proof. This gadget assumes that libff::Fr<nppT> can be
+    /// represented as a single `libff::Fr<wppT>`, and internally asserts this.
+    /// (Expected to be primary inputs to the wrapping statement).
+    std::array<libsnark::pb_variable_array<libff::Fr<wppT>>, NumProofs>
+        nested_primary_inputs;
+
+    /// The array of the results of the verifiers. 1 meaning that the nested
+    /// proof is valid, 0 meaning it may not be valid. (Expected to be a
```

Comment:
0 meaning it may not be valid -> 0 meaning it is not valid

Reply:
I hesitate to write "0 meaning it is not valid".  As discussed before, the gadget will set the value based on the validity of the proof, but the effect on the statement of a 0 here is to simply disable the check, not to actually *prove* that the proof is invalid.

Reply:
Exact, I don't see when this flag would be set to 0 for a valid proof though.
A more accurate description would be the one you said in your comment above, and thus change:
```
    /// The array of the results of the verifiers. 1 meaning that the nested
    /// proof is valid, 0 meaning it may not be valid.
```

to something like:
```
    /// 1 meaning that the nested proof is claimed correct and will be verified, 0 meaning the proof is deemed incorrect and it's verification will be ignored
```

Reply:
As pointed out by @dtebbs offline enough context is already provided around the `result_is_one` variable (see https://github.com/clearmatics/zecale/blob/244ee382e57980992b157a6bc0acb64b8c190766/libzecale/circuits/pairing/bls12_377_pairing.hpp#L467 for instance). As such, no need to repeat ourselves here again.